helpers/stable_swap_math/mod.rs

```diff
@@ -0,0 +1,934 @@
+                .ok_or(MathError::DivByZero(14))?
+                .as_u128();
+            let difference = if ideal_u128 > new_reserves[i] {
+                ideal_u128
+                    .checked_sub(new_reserves[i])
+                    .ok_or(MathError::SubUnderflow(25))?
+            } else {
+                new_reserves[i]
+                    .checked_sub(ideal_u128)
+                    .ok_or(MathError::SubUnderflow(26))?
+            };
+            let fee = _fees.normalized_trade_fee(n_coins as u32, difference)?;
+            // new_u128 is for calculation D2, the one with fee charged
+            new_reserves[i] = new_reserves[i]
+                .checked_sub(fee)
+                .ok_or(MathError::SubUnderflow(27))?;
+        }
+        let d_2 = compute_d(&new_reserves, amp_coef)?;
+        // d0 > d1 > d2,
+        // (d0-d2) => burn_shares (plus fee),
+        // (d0-d1) => diff_shares (without fee),
+        // (d1-d2) => fee part,
+        // burn_shares = diff_shares + fee part
+
+        let burn_shares = U256::from(pool_token_supply)
+            .checked_mul(d_0.checked_sub(d_2).ok_or(MathError::SubUnderflow(28))?)
+            .ok_or(MathError::MulOverflow(23))?
+            .checked_div(d_0)
+            .ok_or(MathError::DivByZero(15))?
+            .as_u128();
+        let diff_shares = U256::from(pool_token_supply)
+            .checked_mul(d_0.checked_sub(d_1).ok_or(MathError::SubUnderflow(29))?)
+            .ok_or(MathError::MulOverflow(24))?
+            .checked_div(d_0)
+            .ok_or(MathError::DivByZero(16))?
+            .as_u128();
+
+        Ok((
+            burn_shares,
+            burn_shares
+                .checked_sub(diff_shares)
+                .ok_or(MathError::SubUnderflow(30))?,
+        ))
+    } else {
+        let burn_shares = U256::from(pool_token_supply)
+            .checked_mul(d_0.checked_sub(d_1).ok_or(MathError::SubUnderflow(31))?)
+            .ok_or(MathError::MulOverflow(25))?
+            .checked_div(d_0)
+            .ok_or(MathError::DivByZero(17))?
+            .as_u128();
+        Ok((burn_shares, 0))
+    }
+}
+
+pub fn rated_compute_lp_amount_for_withdraw(
+    rates: &[u128],
+    withdraw_amounts: &[u128],
+    old_reserves: &Vec<u128>,
+    pool_token_supply: u128,
+    fees: Option<&Fees>,
+    amp_coef: u128,
+) -> Result<(u128, u128), MathError> {
+    let r_withdraw_amounts = amounts_to_rated(withdraw_amounts, rates)?;
+    let r_old_reserves = amounts_to_rated(old_reserves, rates)?;
+
+    compute_lp_amount_for_withdraw(
+        &r_withdraw_amounts,
+        &r_old_reserves,
+        pool_token_supply,
+        fees,
+        amp_coef,
+    )
+}
+
+/// compute amounts to withdraw for lp tokens (no fee)
+/// returns <amounts_to_withdraw, new_reserves>
+pub fn compute_withdraw_amounts_for_lp(
+    lp_amount: u128,
+    old_reserves: &Vec<u128>,
+    pool_token_supply: u128,
+) -> Result<Vec<u128>, MathError> {
+    let mut amounts = Vec::new();
```

Comment:
replace with
```rust
let mut amounts = vec![0; old_reserves.len()];
```
and then, instead of `push`, use index-based access:
```rust
amounts[i] = U256::from(old_reserves[i])
 // ....
```

Reply:
[Various small things](https://github.com/Cardinal-Cryptography/common-amm-stable-swap/pull/5/commits/77337f17f3b8eeb2accacb62c0c60d728ecb9814)